utils/filesystem/files.go

```diff
@@ -1129,6 +1128,60 @@ func (fs *VFS) LsWithExclusionPatterns(dir string, exclusionPatterns ...string)
 	return
 }
 
+func (fs *VFS) LsRecursive(ctx context.Context, dir string) (files []string, err error) {
+	limits := NoLimits()
+	return fs.LsRecursiveWithExtensionPatternsAndLimits(ctx, dir, limits)
+}
+
+func (fs *VFS) LsRecursiveWithExtensionPatterns(ctx context.Context, dir string, exclusionPatterns ...string) (files []string, err error) {
+	limits := NoLimits()
+	return fs.LsRecursiveWithExtensionPatternsAndLimits(ctx, dir, limits, exclusionPatterns...)
+}
+
+func (fs *VFS) LsRecursiveWithExtensionPatternsAndLimits(ctx context.Context, dir string, limits ILimits, exclusionPatterns ...string) (files []string, err error) {
+	err = parallelisation.DetermineContextError(ctx)
+	if err != nil {
+		return
+	}
+	if limits == nil {
+		err = fmt.Errorf("%w: missing file system limits definition", commonerrors.ErrUndefined)
+		return
+	}
+	exclusionRegex, err := NewExclusionRegexList(fs.PathSeparator(), exclusionPatterns...)
+	if err != nil {
+		return
+	}
+	if IsPathExcluded(dir, exclusionRegex...) {
+		// In this case, the whole tree is excluded and so, we can stop here
+		return
+	}
+
+	fn := func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		relativePath := strings.TrimPrefix(path, dir)
+		currentDepth := int64(strings.Count(relativePath, string(fs.PathSeparator())))
+
+		if limits.Apply() && currentDepth > limits.GetMaxDepth() {
+			return filepath.SkipDir
+		}
+
+		if limits.Apply() && int64(len(files)) >= limits.GetMaxFileCount() {
+			return filepath.SkipAll
```

Comment:
If we look at https://github.com/ARM-software/golang-utils/blob/427207db9c482d6bfe81ede28eb30c5338665ed6/utils/filesystem/files.go#L143 then we can see that if you do filepath.SkipAll then it will return the error err=filepath.SkipAll from the parent function since it isn't handled in the same way as SkipDir.

We want to make sure that any errors that are returned from golang-utils are wrapped in our commonerrors so this one should probably return something like `commonerrors.ErrTooLarge` and wrapped with a nice error message saying the number of files exceeds the limit